aptos-move/block-executor/src/scheduler.rs

```diff
@@ -358,100 +470,92 @@ impl Scheduler {
             return None;
         }
 
-        let mut status = self.txn_status[txn_idx].lock();
-        if let TransactionStatus::ReadyToExecute(incarnation, maybe_condvar) = &*status {
+        // Note: we could upgradable read, then upgrade and write. Similar for other places.
+        // However, it is likely an overkill (and overhead to actually upgrade),
+        // while unlikely there would be much contention on a specific index lock.
+        let mut status = self.txn_status[txn_idx].0.write();
+        if let ExecutionStatus::ReadyToExecute(incarnation, maybe_condvar) = &*status {
             let ret = (*incarnation, maybe_condvar.clone());
-            *status = TransactionStatus::Executing(*incarnation);
+            *status = ExecutionStatus::Executing(*incarnation);
             Some(ret)
         } else {
             None
         }
     }
 
     /// If the status of transaction is Executed(incarnation), returns Some(incarnation),
-    /// otherwise returns None. Useful to determine when a transaction can be validated,
-    /// and to avoid a race in dependency resolution.
-    fn is_executed(&self, txn_idx: TxnIndex) -> Option<Incarnation> {
+    /// Useful to determine when a transaction can be validated, and to avoid a race in
+    /// dependency resolution.
+    /// If allow_committed is true, then committed transaction is also considered executed
+    /// (for dependency resolution purposes). If allow_committed is false, then we are
+    /// checking if a transaction may be validated, and a committed (in between) txn does
+    /// not need to be scheduled for validation - so can return None.
+    fn is_executed(&self, txn_idx: TxnIndex, allow_committed: bool) -> Option<Incarnation> {
```

Comment:
this allow_committed comment is confusing, what I understand is that false for validation idx, true for dependencies?

Reply:
updated